### PR TITLE
Add SQL code analysis from PackageReference or ProjectReference

### LINKED ISSUE #332: Add code analyzers support as nuget-packages to new SDK
**Is your feature request related to a problem? Please describe.**
We aim to control code quality in SQL projects with common (among other programming languages) set of rules related to: formatting, naming, bad practices detection. We produce our own linter which has a CLI app and is implemented as extension to VS 2019. But both of ways are divided from the build/compile/code analysis processes occurred during development process. 

However there is a possibility to develop a _Custom Static Code Analysis Rule Assembly_ which is supported by SSDT described in [this msdn article](https://learn.microsoft.com/en-us/sql/ssdt/walkthrough-author-custom-static-code-analysis-rule-assembly?view=sql-server-ver16). But the only way to install this custom code analyzer is to copy it manually somewhere in the deep of `Program Files` folder. This is very uncomfortable anyways and is a **no go** in corporate world where one does not have administrative privileges on his computer and just **cannot** modify anything inside `Program Files`. Even if one managed to install this analyzer once - how would _all the users_ of this analyzer get updates, new releases?

Also currently you still need to enable custom analyzer manually in VS and this checkbox is not a part of project sources or sqlproj PropertyGroups thus you have no idea whether all the developer use this analyzer or not, it did not give you any warnings because there are no warnings or just because you forgot to enable this analyzer. This makes mentioned solution also _unreliable_.

**Describe the solution you'd like**
New SDK could just import all the nuget-packages which are marked somehow as code analyzers on the fly and apply them. Just as Roselyn does.

Scenario of using StyleCop code analyzer in C# development process:
- add nuget dependency from StyleCop to your C# project
- commit
- clone repo
- build your project
- there you go - nuget package was restored, SDK detected code analyzer and applied it during build process
- you see Warnings, Errors and Info messages in the dedicated VS tab

It'd be great if you make SDK:
- import _all_ analyzers from nuget packages - a project might require more than one analyzer
- implement these analyzers support in VS 2019, VS 2022 and VS Code

**Describe alternatives you've considered**
If this will not be implemented then perhaps we will try this scenario:
- deliver our linter as a nuget
- add custom target to all sqlproj of ours which will call this linter during build process
but error messages will not be integrated into VS IDE if I'm not mistaken.

**Additional context**
- [StyleCop nuget](https://www.nuget.org/packages/StyleCop.Analyzers/1.2.0-beta.507#readme-body-tab)
- [SSDT custom analyzer development guide](https://learn.microsoft.com/en-us/sql/ssdt/walkthrough-author-custom-static-code-analysis-rule-assembly?view=sql-server-ver16)

### PR DESCRIPTION
**Pull request by zijchen**
Corresponding SDK changes for analyzer package support, fixes #332 

TODO: Need new DacFx version

Add PackageReference to analyzer Nuget via
```xml
<PackageReference Include="CodeAnalyzerSample" Version="1.0.0" />
```
* Analyzer DLL must be in analyzers/dotnet/cs
* Analyzer package can target any .NET version DacFx supports

You can also add reference via ProjectReference
```xml
<ProjectReference Include="C:\test\test_analyzer\CodeAnalyzerSample.csproj"
                      PrivateAssets="all"
                      ReferenceOutputAssembly="false"
                      OutputItemType="Analyzer"
                      SetTargetFramework="TargetFramework=netstandard2.1" />
```
* You must override target `GetTargetPath` in your csproj

See example analyzer in CodeAnalyzerSample.csproj

### CODE REVIEW

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    <TargetFrameworks>netstandard2.1</TargetFrameworks>
+    <TargetsForTfmSpecificContentInPackage>$(TargetsForTfmSpecificContentInPackage);_AddAnalyzersToOutput</TargetsForTfmSpecificContentInPackage>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.SqlServer.DacFx" Version="162.3.566" PrivateAssets="all" />
+  </ItemGroup>
+  <Target Name="GetTargetPath" />
+  <Target Name="_AddAnalyzersToOutput">
+    <ItemGroup>
+      <TfmSpecificPackageFile Include="$(OutputPath)\CodeAnalyzerSample.dll" PackagePath="analyzers/dotnet/cs" />
```
Comment (ErikEJ) [inline test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:13]: Why this uncommon location? How about "tools" ?

**Comment by ErikEJ** [comment]
This is very cool, looking forward. 

I already publish a couple of NuGet packages with analyzer rules, targeting both .net standard 2.1 and netfx 4.6.2: https://www.nuget.org/packages?q=erikej.dacfx

So very interested in this topic.

The packages that I publish are currently used for consumption by a project, that use DacFX, not for consumption in a .sqlproj project. But it would be very nice to be able to support both scenarios in the same NuGet package, IMHO.

**Comment by ErikEJ** [comment]
What does "You must override target GetTargetPath in your csproj" mean??

**Comment by ErikEJ** [comment]
So, this is the result of Nuget pack with the sample here:

image

And this is my package:

image

Questions:

What is the value of having the .dll in two places?

How do I target multiple frameworks? Or is this a .NET only proposal? (No NetFX) ?

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj (hunk 1 above)
Comment (zijchen) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:13]: This is the [path Nuget natively resolves analyzers from](https://learn.microsoft.com/nuget/guides/analyzers-conventions#analyzers-path-format)

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (zijchen) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:1]: I couldn't find a straightforward sample either, seems like .NET doesn't have an easy path to creating an analyzer package. The sample I used is from https://github.com/dotnet/samples/blob/main/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Package/MakeConst.Package.csproj

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj (hunk 1 above)
Comment (ErikEJ) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:13]: Ok, but this is not a roslyn analyzer and the target language is not "cs" ...

**Comment by zijchen** [comment]
> What does "You must override target GetTargetPath in your csproj" mean??

This is only necessary if you want to add the analyzer via ProjectReference. It's to workaround a bug of a csproj referenced from sqlproj throwing that the target doesn't exist
https://github.com/microsoft/DacFx/blob/79da90409dd6116b9dd902a0d5b0bf908a38b4b8/test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj#L10

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ErikEJ) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:1]: Again, this is not a roslyn analyzer and DacFx has its own analysis result mechanism. 

But OK  - there will be few publishers of rules and hopefully many consumers

**Comment by ErikEJ** [comment]
Is it really necessary ti support that?

**Comment by zijchen** [comment]
> This is very cool, looking forward.
> 
> I already publish a couple of NuGet packages with analyzer rules, targeting both .net standard 2.1 and netfx 4.6.2: https://www.nuget.org/packages?q=erikej.dacfx
> 
> So very interested in this topic.
> 
> The packages that I publish are currently used for consumption by a project, that use DacFX, not for consumption in a .sqlproj project. But it would be very nice to be able to support both scenarios in the same NuGet package, IMHO.

We are introducing a new property `SqlCodeAnalysisAssemblyPaths` in the next version of DacFx that you can probably use to hook up additional analyzer paths. I imagine you can use this for MSBuild.Sdk.Sqlproj too

**Comment by ErikEJ** [comment]
Amazing!

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (zijchen) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:1]: We can add this in a future `dotnet new` template @dzsquared

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj (hunk 1 above)
Comment (zijchen) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:13]: Yeah I can understand that, my motivation was to keep it similar to C# and use Nuget's package resolution logic. We did build in the extensibility via `SqlCodeAnalysisAssemblyPaths` so if there is a growing demand for another location, we can easily hook that up.

Comment (ErikEJ) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/CodeAnalyzerSample.csproj:13]: > Nuget's package resolution logic

Fair, if that adds value for you.

But maybe just use `analyzers\dotnet\` ??

**Comment by jmezach** [comment]
@ErikEJ That would make it a lot easier to implement https://github.com/rr-wfm/MSBuild.Sdk.SqlProj/issues/569.

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/TableNameEndingInViewRule.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+            TSqlObject table = ruleExecutionContext.ModelElement;
+            if (table != null)
+            {
+                if (NameEndsInView(table.Name))
+                {
+                    string problemDescription = string.Format("Table name {0} ends in View. This can cause confusion and should be avoided",
+                                                              GetQualifiedTableName(table.Name));
+                    SqlRuleProblem problem = new SqlRuleProblem(problemDescription, table);
+                    //SqlRuleProblem problem = new SqlRuleProblem(this, problemDescription, table);
```
Comment (chlafreniere) [inline test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/TableNameEndingInViewRule.cs:47]: Can this be removed?

**Comment by ErikEJ** [comment]
@zijchen I can see there is a preview release now with `SqlCodeAnalysisAssemblyPaths` I assume

In which namespace do I find `SqlCodeAnalysisAssemblyPaths` ?

**Comment by zijchen** [comment]
> @zijchen I can see there is a preview release now with `SqlCodeAnalysisAssemblyPaths` I assume
> 
> In which namespace do I find `SqlCodeAnalysisAssemblyPaths` ?

`SqlCodeAnalysisAssemblyPaths` is a property of `Microsoft.Data.Tools.Schema.Tasks.Sql.SqlStaticCodeAnalysisTask`. This hooks into a new `AssemblyLookupPath` property in `Microsoft.SqlServer.Dac.CodeAnalysis.CodeAnalysisServiceSettings` if you're using the library directly.

File: test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/TableNameEndingInViewRule.cs (hunk 3 above)
Comment (zijchen) [reply test/Microsoft.Build.Sql.Tests/TestData/CodeAnalyzerSample/TableNameEndingInViewRule.cs]: Removed

**Review by llali — APPROVED** [review]

**Comment by ErikEJ** [comment]
@zijchen I actually  just did this to pack my rules:

```xml
    <None Include="..\SqlServer.Dac\bin\$(Configuration)\netstandard2.1\SqlServer.Dac.dll" 
             Pack="true"
             PackagePath="analyzers\dotnet\cs"
             Visible="false" />
    <None Include="..\SqlServer.Rules\bin\$(Configuration)\netstandard2.1\SqlServer.Rules.dll" 
        Pack="true"
        PackagePath="analyzers\dotnet\cs"
        Visible="false" />

```

**Comment by ErikEJ** [comment]
Sadly, the property in DacFX ended up being called: `SqlCodeAnalysisAssemblyPath` 😢  - `s` is missing

**Pull request merged**